### PR TITLE
try local setter for options.vnode

### PR DESCRIPTION
**Pull request by JoviDeCroock**

### CODE REVIEW

File: src/options.js [hunk 1]
```diff
@@ -13,4 +15,21 @@ const options = {
...
+		return vnodeHook;
+	},
+	set(newHook) {
+		if (newHook) {
+			let oldHook = vnodeHook;
+			vnodeHook = (vnode) => {
+				newHook(vnode);
+				if (oldHook) oldHook(vnode);
+			};
+		} else {
```
Comment (JoviDeCroock) [inline src/options.js:29]: we need to handle undefined so we can reset the value in our tests

File: src/options.js [hunk 2]
```diff
@@ -13,4 +15,21 @@ const options = {
...
+Object.defineProperty(options, 'vnode', {
+	get() {
+		return vnodeHook;
+	},
+	set(newHook) {
+		if (newHook) {
+			let oldHook = vnodeHook;
+			vnodeHook = (vnode) => {
+				newHook(vnode);
+				if (oldHook) oldHook(vnode);
```
Comment (andrewiggins) [inline src/options.js:27]: Yes!

Though I think we can't auto apply the old hook cuz I think some implementations need to invoke their logic before and/or after the previous hooks. For example, if I want to modify something `preact/hooks` does I may want to invoke my implementation after invoking the previous hook to ensure hooks stuff runs first.

However, I like this idea of if simplifying hook integrations. I think we had previously discussed adding something like a `preact/shared-internals` package or something and exposing a [hook helper like we have in signals](https://github.com/preactjs/signals/blob/6b6af05d433b4d2a8da0be5a65ba5305d49e7b73/packages/preact/src/index.ts#L27-L31) that would always ensure the first argument to the hook is the previous hook. That doesn't need to happen in this PR but just thought I'd write it down cuz I'm not sure we have that written down elsewhere 😅

Comment (JoviDeCroock) [reply src/options.js]: I updated it, just have one test that seems to be failing, not entirely sure how to remedy it yet will look deeper over the next week.

**Comment by marvinhagemeister** [comment]
I like this PR a lot!

File: src/options.js (hunk 2 above)
Comment (JoviDeCroock) [reply src/options.js]: The perf diff does not seem to be working atm

File: src/options.js [hunk 3]
```diff
@@ -9,8 +21,105 @@ import { _catchError } from './diff/catch-error';
  * ctrl+click or cmd+click on mac the type definition below).
  * @type {import('./internal').Options}
  */
-const options = {
-	_catchError
-};
+const options = {};
+
+Object.defineProperty(options, 'vnode', {
```
Comment (developit) [inline src/options.js:26]: I wonder if we could use a helper to define the value field and the options hook property at the same time?

```js
export let vnode = optionsHook('vnode');

/**
 * @template {keyof import('./internal').Options} T
 * @param {T} name
 * @param {any} [value] - initial value for the hook
 * @returns {import('./internal').Options[T]}
 */
function optionsHook(name, value) {
	Object.defineProperty(options, 'unmount', {
		get() {
			return value;
		},
		set(newValue) {
			value = newValue;
		}
	});
	return value;
}
```

**Comment by JoviDeCroock** [comment]
Something weird is going on here, locally the tests fail however in CI they are good 😅 could be a diff between watch and normal runs

File: src/options.js (hunk 3 above)
Comment (JoviDeCroock) [reply src/options.js]: @developit the byte-size change is actually very minimal

**Comment by JoviDeCroock** [comment]
Closing as the issues with tests are a bit too much

**Pull request closed without merging**